src/maps/hashmap.js

```diff
@@ -0,0 +1,134 @@
+const IAMap = require('iamap')
+const murmurhash3 = require('murmurhash3js-revisited')
+const Type = require('../base')
+const Block = require('@ipld/block')
+
+const _type = 'IPLD/Experimental/HashMap/0'
+
+class HashMap extends Type {
+  get _type () {
+    return _type
+  }
+
+  get kind () {
+    return 'map'
+  }
+
+  async get (args, continuation) {
+    let path = args.path.split('/').filter(Boolean)
+    let key = path.shift()
+
+    if (!key) {
+      throw new Error('HashMap type needs a string to look up')
+    }
+
+    // TODO: this should be a validation of this.node against the HAMT root block schema
+    if (!IAMap.isRootSerializable(this.node) || !IAMap.isSerializable(this.node)) {
+      throw new Error('Block does not contain expected HashMap root data')
+    }
+
+    // provide a store to IAMap that won't do any retrieval but will intercept load() calls
+    // so we can pass that up for continuation
+    let target
+    const fauxStorage = {
+      save (obj) {
+        throw new Error('Mutations are not supported during get() operations')
+      },
+      load (id) {
+        if (target) {
+          throw new Error('Single load() expected during traversal')
+        }
+        target = id
+        return null
+      },
+      isEqual (cid1, cid2) {
+        return cid1.equals(cid2)
+      }
+    }
+
+    // instantiate a root IAMap node
+    let root = IAMap.fromSerializable(fauxStorage, null, this.node)
+    let map = root
+    let depth = 0
+    if (continuation) {
+      // we have navigated to a child node of with depth of continuation.depth
+      if (!IAMap.isSerializable(continuation.response.node)) {
+        throw new Error('Block does not contain expected HashMap child node data')
+      }
+      depth = continuation.depth + 1
+      // instantiate a child node based on the config of the root node but at `depth` depth
+      map = root.fromChildSerializable(null, continuation.response.node, depth)
+    }
+
+    try {
+      let result = await map.get(key)
+      if (!path.length) {
+        return { result }
+      }
+      throw new Error('Traversing into values is not yet supported')
+      // what do? could be an arbitrary kind
+      // return { call: { info: { method: 'get', args: path.join('/') }, target: result, proxy: true } }
+    } catch (err) {
+      if (!target) {
+        throw new Error(`No such element in map: ${key} (${err.message})`)
+      }
+
+      // having a `target` means that IAMap made a `load() call from storage, let's go fetch what it wants
+      let info = { continuation: { depth }, method: 'get', args: { path: '/' } }
+      let call = { info, target }
+      return { call }
+    }
+  }
+}
+
+HashMap._type = _type
+
+// register an x86 32-bit murmurhash as available for use and make it the default (in create() below)
+// this is probably not what we want to end up doing
```

Comment:
here's something for you to ponder and connects to a discussion in https://github.com/ipld/specs/pull/109 about HAMT properties—you should be able to configure what hash function you're using, but we can't support every conceivable hash that might be usable so I've built in a `registerHasher()` mechanism to make it fully pluggable.
But having that much flexibility gets in the way of the transparent navigation here, a user shouldn't have to register basic hashes that are commonly in use.

So I'm thinking that re HAMT spec:

1. hash function should be configurable
2. HAMT (or "HashMap") implementaitons should ship with a good fast hash and a good cryptographic hash by default, maybe murmur-x64-128 and sha2-256
3. if a HAMT you want to navigate through uses some alternate hash, there should be some way to register that through this type system

that registering process isn't likely to be unique to the HAMT either, encryption is going to have all sorts of flexibility but I can imagine other collections having configurability that we may want to leave pluggable to allow for the user to make it work best for their use-case which may be considerably different to what we design for.

Here I'm just registering murmur-x86-32 for simplicity.

Reply:
We have a similar issue in the `unixfsv2` spec (once I update it) in that we want to say “This property can be any Binary Type” but we obviously can’t ensure compatibility if we don’t have a list of required types to be supported.

The solution I’ve been kicking around in my head is, each version of spec has a list of required multi-block types. If we implement better multi-block types in the future then we put out a new version of the spec noting the new type requirements.

We should do something similar here, the spec should require support for `murmur-x64-128` but the implementation can leave the door open to additional algorithms in future versions.

Reply:
I'd just like us to keep in mind JS API considerations for how to register / side-load alternative hash algorithms for users that may want to do that. It'd either be similar to the `Lookup` mechanism you have or just some global you could hook in to if you absolutely need to. Right now it's `require('iamap').registerHasher(...)` and it'd be registered globally, but it'd be good to have that integrated as some kind of side loading mechanism.

Reply:
If you make it configurable on the class instead of the module then it would match up nicely with the existing Lookup mechanism since you pass in a specific class.

Reply:
The problem is: how do we expose the class? But I guess the answer to that depends on the what the writer interface will end up looking like. Currently that's these hacky `create()` factories but we need something nicer. Something that extends a `NodeBuilder` like in Go would make sense. It'd be really nice to have a base that does all of the saving of blocks and even build in transactional semantics to make mutations cheaper.
Something to defer for now. Perhaps revisit when we have more complex multiblock data structures with weird many-block mutation semantics.